base/dcp_feed_type.go

```diff
@@ -12,14 +16,86 @@ import (
 // Cannot use the serverContext to retrieve this information, as the cbgt manager for a database is initialized
 // before the database is added to the server context's set of databases
 var cbgtCredentials map[string]cbgtCreds
+
+// cbgtBucketToDBName maps bucket names to DB names for DBs currently registered with CBGT. Necessary because in some
+// contexts we only have access to the bucket name.
+var cbgtBucketToDBName map[string]string
+
+// cbgtRootCertPools is a map of bucket UUIDs to cert pools for its CA certs. The documentation comment of
+// cbgtRootCAsProvider describes the behaviour of different values.
+var cbgtRootCertPools map[string]*x509.CertPool
 var cbgtCredentialsLock sync.Mutex
 
 type cbgtCreds struct {
 	username string
 	password string
 }
 
-const SOURCE_GOCOUCHBASE_DCP_SG = "couchbase-dcp-sg"
+const (
+	SOURCE_GOCOUCHBASE_DCP_SG = "couchbase-dcp-sg"
+	SOURCE_GOCB_DCP_SG        = "gocb-dcp-sg"
+)
+
+// cbgtRootCAsProvider implements cbgt.RootCAsProvider. It returns a x509.CertPool factory with the root certificates
+// for the given bucket. Edge cases:
+// * If it returns a function that returns nil, TLS is used but certificate validation is disabled.
+// * If it returns a nil function, TLS is disabled altogether.
+func cbgtRootCAsProvider(bucketName, bucketUUID string) func() *x509.CertPool {
+	if pool, ok := cbgtRootCertPools[bucketUUID]; ok {
```

Comment:
We lock when we write to this map, but not when we read from it here. Should we be doing?

Reply:
Good catch